### ISSUE TITLE
Boost fails to build on ANL's Theta

### ISSUE THREAD

**Issue by dorier** [spack/spack#9613]
I'm trying to install boost with spack on ANL's Theta machine, using the GNU compilers. I tried version 1.68.0 and 1.67.0. In both cases the build fails when calling `b2 --clean`. When I go in the the build directory and call `./b2 -- clean` myself, I get `Illegal instruction (core dumped)`.

The same happens with Intel compiler instead of GNU.

If I try to build boost manually (i.e. downloading the archive, running `./bootstrap.sh` with the libraries I want, then `./b2` and `./b2 install`) it works fine.

### Steps to reproduce the issue

```console
$ spack install boost
==> diffutils is already installed in /gpfs/mira-home/mdorier/spack/opt/spack/cray-cnl6-mic_knl/gcc-7.3.0/diffutils-3.6-hdksfzqzgtfwvwbvhu7s75u6f7tkudvz
==> bzip2 is already installed in /gpfs/mira-home/mdorier/spack/opt/spack/cray-cnl6-mic_knl/gcc-7.3.0/bzip2-1.0.6-m22cmmtaf36fnfmvwy36dssvltqfw73q
==> zlib is already installed in /gpfs/mira-home/mdorier/spack/opt/spack/cray-cnl6-mic_knl/gcc-7.3.0/zlib-1.2.11-yuj7ryis72kd6iwpq3qmvp627umsbymc
==> Installing boost
==> Fetching https://dl.bintray.com/boostorg/release/1.67.0/source/boost_1_67_0.tar.bz2
######################################################################## 100.0%
==> Staging archive: /gpfs/mira-home/mdorier/spack/var/spack/stage/boost-1.67.0-a5gxl76fce2fobsekfnhvrhsl5j6j6cw/boost_1_67_0.tar.bz2
==> Created stage in /gpfs/mira-home/mdorier/spack/var/spack/stage/boost-1.67.0-a5gxl76fce2fobsekfnhvrhsl5j6j6cw
==> Applied patch python_jam.patch
==> Building boost [Package]
==> Executing phase: 'install'
==> Error: ProcessError: Command exited with status -4:
    './b2' '--clean'
```

### Platform and user environment

```commandline
$ uname -a 
Linux thetalogin6 4.4.103-6.38-default #1 SMP Mon Dec 25 20:44:33 UTC 2017 (e4b9067) x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -d
Description:	SUSE Linux Enterprise Server 12 SP3
``` 

- `compilers.yaml`

```
compilers:
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules:
    - PrgEnv-cray
    - cce/8.7.3
    operating_system: cnl6
    paths:
      cc: cc
      cxx: CC
      f77: ftn
      fc: ftn
    spec: cce@8.7.3
    target: any
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules:
    - PrgEnv-gnu
    - gcc/7.3.0
    operating_system: cnl6
    paths:
      cc: cc
      cxx: CC
      f77: ftn
      fc: ftn
    spec: gcc@7.3.0
    target: any
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: cnl6
    paths:
      cc: /opt/cray/pe/craype/2.5.15/bin/cc
      cxx: /opt/cray/pe/craype/2.5.15/bin/CC
      f77: /opt/cray/pe/craype/2.5.15/bin/ftn
      fc: /opt/cray/pe/craype/2.5.15/bin/ftn
    spec: intel@18.0.2.199
    target: any
```

**Comment by dorier** [comment]
I get a similar problem when trying to build `cmake`, which builds fine manual, but not with spack. I think what happens is both cmake and boost are first compiling an executable that is later called during the build process, but on Theta compilers are actually cross-compiling for the compute nodes (cray-cnl6-mic_knl architecture), so some instructions don't exist on the login node where spack is used.

**Comment by jgalarowicz** [comment]
@dorier I just ran into this (boost failure) too!
Thanks for posting.

**Comment by jgalarowicz** [comment]
@dorier Thanks for the information!

**Comment by mwkrentel** [comment]
Fixed in PR #10295.
